### DISCUSSION TITLE
call server component in different component(client, server)

### DISCUSSION THREAD

**Discussion by d1d4r** [vercel/next.js#64828] [Help] [answered]
### Summary

hi there

i have a component named Todo that fetch data from API

```
import axios from "axios";
import React from "react";

export default async function Todo({passProps}) {
  const res = await axios.get(process.env.BASE_URL_TODO);

  return (
    <div>
      <p>{passProps }</p>
      <ul>
        {res.data.todos.map((todo) => {
          return <li key={todo.id}>{JSON.stringify(todo)}</li>;
        })}
      </ul>
    </div>
  );
}
```
and i use it in two different component

one is client

```
"use client";
import React from "react";
import Todo from "./Todo";


export default function ClientTodo({ testProps }) {
  return (
    <div>
      <p>{testProps}</p>
      <Todo passProps="from client" />
    </div>
  );
}
```

another on is server component
```
import React from "react";
import Todo from "./Todo";


export default function ServerTodo({ testProps }) {
  return (
    <div>
      <p>{testProps}</p>
      <Todo passProps="from server" />
    </div>
  );
}

```
and call both in /todo route

```
import ClientTodo from "./components/ClientTodo";
import ServerTodo from "./components/ServerTodo";

export default function DashbordPage({ searchParams }) {
  return (
    <div className="grid grid-rows-1 grid-cols-2 ">
      <ClientTodo testProps="test props to client" />
      <ServerTodo testProps="test props to server" />
    </div>
  );
}
```

when i visit todo route show this error message
![nextjsError](https://github.com/vercel/next.js/assets/68716408/0fc33a5f-9980-44f2-970f-9028d3a4d2d0)

but when i replace
 this code const res = await axios.get(process.env.BASE_URL_TODO);
to   const res = await axios.get('https://dummyjson.com/todos');

every thing work well

why this error happen, if i can't user server component in client component why i can render it when i use this code
const res = await axios.get('https://dummyjson.com/todos');

**Comment by mutasim77** [comment, marked as answer]
The issue here lies with the Environment Variables you're utilizing. Specifically, consider `process.env.BASE_URL_TODO`. It operates within a [node](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#loading-environment-variables) environment (which is server-side; if you use it solely within a server component, it's completely fine). On the other hand, the client component (with 'use client') seeks these variables at the browser level. Therefore, if you want to access them in the browser environment, simply add the prefix ["NEXT_PUBLIC_"](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser). In your case, just modify it to:
 ```js 
const res = await axios.get(process.env.NEXT_PUBLIC_BASE_URL_TODO)
```
 and ofc, make sure to also update it in the `.env` file.

**Reply by d1d4r** [reply]
> The issue here lies with the Environment Variables you're utilizing. Specifically, consider `process.env.BASE_URL_TODO`. It operates within a [node](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#loading-environment-variables) environment (which is server-side; if you use it solely within a server component, it's completely fine). On the other hand, the client component (with 'use client') seeks these variables at the browser level. Therefore, if you want to access them in the browser environment, simply add the prefix ["NEXT_PUBLIC_"](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser). In your case, just modify it to:
> 
>  ```js 
> 
> const res = await axios.get(process.env.NEXT_PUBLIC_BASE_URL_TODO)
> 
> ```
> 
>  and ofc, make sure to also update it in the `.env` file.

thank u bro💙, after asking i found the answer directly in nextjs documentation😅
